var/spack/repos/builtin/packages/ginkgo/package.py

```diff
@@ -63,18 +69,39 @@ class Ginkgo(CMakePackage, CudaPackage, ROCmPackage):
     patch('1.4.0_skip_invalid_smoke_tests.patch', when='@master')
     patch('1.4.0_skip_invalid_smoke_tests.patch', when='@1.4.0')
 
+    # Newer DPC++ compilers use the updated SYCL 2020 standard which change
+    # kernel attribute propagation rules. This doesn't work well with the
+    # initial Ginkgo oneAPI support.
+    patch('1.4.0_dpcpp_use_old_standard.patch', when='+oneapi @master')
+    patch('1.4.0_dpcpp_use_old_standard.patch', when='+oneapi @1.4.0')
+
+    def setup_build_environment(self, env):
+        spec = self.spec
+        if '+oneapi' in spec:
+            env.set('MKLROOT',
+                    join_path(spec['intel-oneapi-mkl'].prefix,
+                              'mkl', 'latest'))
+            env.set('DPL_ROOT',
+                    join_path(spec['intel-oneapi-dpl'].prefix,
+                              'dpl', 'latest'))
+
     def cmake_args(self):
         # Check that the have the correct C++ standard is available
         if self.spec.satisfies('@:1.2.0'):
             try:
                 self.compiler.cxx11_flag
             except UnsupportedCompilerFlag:
-                InstallError('Ginkgo requires a C++11-compliant C++ compiler')
+                raise InstallError('Ginkgo requires a C++11-compliant C++ compiler')
         else:
             try:
                 self.compiler.cxx14_flag
             except UnsupportedCompilerFlag:
-                InstallError('Ginkgo requires a C++14-compliant C++ compiler')
+                raise InstallError('Ginkgo requires a C++14-compliant C++ compiler')
+
+        cxx_is_dpcpp = os.path.basename(self.compiler.cxx) == "dpcpp"
```

Comment:
This check also appears to work if I do the following:

```
spack install intel-oneapi-compilers              # in my case I didn't have a system version of the oneAPI compilers
spack compiler add `spack location -i intel-oneapi-compilers`/compiler/latest/linux/bin
```

And then edit the new entry as follows (there are 2 changed lines, each ending with `#`)

```
- compiler:
    spec: oneapi@2021.4.0-dpcpp # changed the name
    paths:
      cc: /.../spack/opt/spack/linux-ubuntu20.04-haswell/gcc-9.3.0/intel-oneapi-compilers-2021.4.0-<hash>/compiler/latest/linux/bin/icx
      cxx: /.../spack/opt/spack/linux-ubuntu20.04-haswell/gcc-9.3.0/intel-oneapi-compilers-2021.4.0-<hash>/compiler/latest/linux/bin/dpcpp # changed from icpx
      f77: /.../spack/opt/spack/linux-ubuntu20.04-haswell/gcc-9.3.0/intel-oneapi-compilers-2021.4.0-<hash>/compiler/latest/linux/bin/ifx
      fc: /.../spack/opt/spack/linux-ubuntu20.04-haswell/gcc-9.3.0/intel-oneapi-compilers-2021.4.0-<hash>/compiler/latest/linux/bin/ifx
    flags: {}
    operating_system: ubuntu20.04
    target: x86_64
    modules: []
    environment: {}
    extra_rpaths: []
```

In other words if I do steps 1 and 2 of your described approach. Did you find you also had to do

> In lib/spack/spack/compilers/oneapi.py, replace icpx by dpcpp everywhere
> Copy or link lib/spack/env/oneapi/ipcx to lib/spack/env/oneapi/dpcpp, and add dpcpp to the list of C++ compilers.

? Admittedly things look odd without those changes, but once you change the cxx entry in `compilers.yaml` to `dpcpp`, then Spack should use that (regardless of whether it invokes a symlink called `icpx`).

I'm still looking into making this easier, but I wanted to check if the process you outlined could be simplified as well.

Reply:
Thanks for looking into this already. Indeed it seems like the symlink step works from the script point of view for `dpcpp` detection, but the problem then comes from `oneMKL` where they also have a step to determine the compiler and change options etc depending on what they find. In particular, in `MKLConfig.cmake` this portion doesn't seem to like the `icpx` symlink:
``` CMake
if(CMAKE_CXX_COMPILER MATCHES ".*dpcpp[.exe]*")
  set(DPCPP_COMPILER ON)
endif()
```

If I instead apply the extra steps for creating a new `dpcpp` symlink, using MKL in our project works as expected. Otherwise, it's incorrectly detected (the CMake target `MKL::MKL_DPCPP` which we rely on doesn't exist).

Reply:
A few notes:

* It looks like `MKLConfig.cmake` is provided by MKL (I don't see it in the `ginkgo` source files) so setting `DPCPP_COMPILER=ON` in the `cmake_args` function of `ginkgo` might work but I don't think it would address the problem as directly as it should.
* `icpx` and `dpcpp` both give exactly the same output when you ask for `--version` (so I don't currently see a way to detect a difference other than to test the names)

I added a compiler definition for dpcpp at #27168 such that you should be able to `spack install ginkgo%dpcpp` (after `spack compiler add...`) without any additional steps, but I need to consider whether there is a cleaner approach.

Reply:
Thanks for your dpcpp addition. That seems like a nice approach at first glance.

Yes, the `MKLConfig.cmake` is what the MKL package provides in their CMake config after installing `intel-oneapi-mkl`, and there you find that explicit `DPCPP` portion which I pasted previously. I believe they need this because if the compiler is DPCPP they probably rely on some JIT features of the compiler, which I don't think `icpx` provides?